### PR TITLE
Implement ContextFromTwigRenderMethodCallCollector

### PR DESCRIPTION
**Pull request by VincentLanglet**
Wip @ruudk

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,7 +48,7 @@ Now that we have a flat template, we don't know anything about the context the t
 
 We use PHPStan to run the [BlockContextCollector](src/PHPStan/Collector/BlockContextCollector.php). This collector gathers the context before rendering every block or parent block call.
 
-While running PHPStan, it's also a good time to search for places that render the template. [ContextFromReturnedArrayWithTemplateAttributeCollector](src/PHPStan/Collector/ContextFromReturnedArrayWithTemplateAttributeCollector.php) and [ContextFromRenderMethodCallCollector](src/PHPStan/Collector/ContextFromRenderMethodCallCollector.php) search for controllers that render a Twig template.
+While running PHPStan, it's also a good time to search for places that render the template. [ContextFromReturnedArrayWithTemplateAttributeCollector](src/PHPStan/Collector/ContextFromReturnedArrayWithTemplateAttributeCollector.php) and [ContextFromRenderMethodCallCollector](src/PHPStan/Collector/ContextFromControllerRenderMethodCallCollector.php) search for controllers that render a Twig template.
```
Comment (VincentLanglet) [inline README.md:51]: Todo update the doc

File: src/PHPStan/Collector/ContextCollector.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ *      startLine: int,
+ *      endLine: int,
+ *      template: string,
+ *      context: string,
+ *  }
+ *
+ * @template-covariant TNodeType of Node
+ * @implements Collector<TNodeType, array<TemplateData>>
+ */
+interface ContextCollector extends Collector
```
Comment (VincentLanglet) [inline src/PHPStan/Collector/ContextCollector.php:19]: It's simpler for the implementation if ALL the collector have the same signature. `array<TemplateData>`

File: src/PHPStan/Collector/ContextFromControllerRenderMethodCallCollector.php [hunk 3]
```diff
@@ -54,22 +48,27 @@ public function processNode(Node $node, Scope $scope): ?array
             return null;
         }
 
-        if (count($node->args) !== 2) {
+        if (count($node->args) < 2) {
```
Comment (VincentLanglet) [inline src/PHPStan/Collector/ContextFromControllerRenderMethodCallCollector.php:51]: 3 args is allowed.

I also need to allow 1 args cause
```
$this->render($foo)
```
is valid, and we need to report someone who forgot passing the context for a template.

File: src/PHPStan/Collector/ContextFromControllerRenderMethodCallCollector.php [hunk 4]
```diff
@@ -54,22 +48,27 @@ public function processNode(Node $node, Scope $scope): ?array
...
             return null;
         }
 
-        if (!$node->args[0]->value instanceof Node\Scalar\String_) {
+        $views = $scope->getType($node->args[0]->value)->getConstantStrings();
```
Comment (VincentLanglet) [inline src/PHPStan/Collector/ContextFromControllerRenderMethodCallCollector.php:55]: I'm trying to support
```
if ($bool) {
    $foo = 'template1';
} else {
    $foo = 'template2';
}

$this->render($foo);
```

File: src/PHPStan/Collector/ContextFromTwigRenderMethodCallCollector.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        if (!\in_array($node->name->name, ['render', 'display'], true)) {
+            return null;
+        }
+
+        $varType = $scope->getType($node->var);
+        if (!(new ObjectType(Environment::class))->isSuperTypeOf($varType)->yes()) {
+            return null;
+        }
+
+        if (count($node->args) < 2) {
```
Comment (VincentLanglet) [inline src/PHPStan/Collector/ContextFromTwigRenderMethodCallCollector.php:46]: Same, I need to support 1 param.

**Review by VincentLanglet — COMMENTED** [review]
Some test would be great

**Comment by ruudk** [comment]
Could you rebase this PR against `main`?

You can add a test like below in this location:
```
tests/EndToEnd/RenderPoints/RenderFromAbstractController.php
```

```php
<?php

declare(strict_types=1);

namespace TwigStan\EndToEnd\RenderPoints;

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;

class RenderFromAbstractController extends AbstractController
{
    public function renderAction(): Response
    {
        $response = new Response(status: Response::HTTP_CREATED);

        return $this->render('EndToEnd/RenderPoints/list.html.twig', [
            'title' => 'RenderAction',
            'artists' => ['Adele', 'Kanye West'],
        ], $response);
    }

    public function renderViewAction(): Response
    {
        return new Response($this->renderView('EndToEnd/RenderPoints/list.html.twig', [
            'title' => 'RenderViewAction',
            'artists' => ['Adele', 'Kanye West'],
        ]));
    }
}

```

**Comment by VincentLanglet** [comment]
Test are not working as expected @ruudk 

I assume it's because with such render method the template does not start with `@`.
I'm not sure to understand how was configured
- The AbsolutePathLoader
- The tests twig-loader

Could you take a look ?

**Comment by ruudk** [comment]
@VincentLanglet I found the problem. Pushed a fix. Now there is only 1 error left, which is expected.

**Comment by ruudk** [comment]
Thanks @VincentLanglet

**Pull request merged**